tools/perfdw.d

```diff
@@ -0,0 +1,95 @@
+/*
+ * Trace all IOs for each downstairs from the time they are sent over
+ * the network socket to when the ack for an IO is received.
+ * Group by IO type (R/W/F) and client ID (Which downstairs).
+ *
+ * arg0 is the job ID number.
+ * arg1 is the client ID
+ */
+cdt*:::gw_read_submit_start,
+cdt*:::gw_write_submit_start,
+cdt*:::gw_flush_submit_start
+/arg1 == 0/
+{
+    start0[arg0] = timestamp;
+}
+cdt*:::gw_read_submit_start,
+cdt*:::gw_write_submit_start,
+cdt*:::gw_flush_submit_start
+/arg1 == 1/
+{
+    start1[arg0] = timestamp;
+}
+cdt*:::gw_read_submit_start,
+cdt*:::gw_write_submit_start,
+cdt*:::gw_flush_submit_start
+/arg1 == 2/
+{
+    start2[arg0] = timestamp;
+}
+
+/*
+ * When a read ACK comes back, calculate the delta and store it
+ */
+cdt*:::gw_read_submit_end
+/start0[arg0] != 0 && arg1 == 0/
```

Comment:
Perhaps something like this would work:

```dtrace
cdt*:::gw_*_submit_done
/start[arg0, arg1]/
{
    strtok(probename, "_");
    this->cmd = strtok(NULL, "_");

    @[strjoin(this->cmd, " downstairs client"), arg1] = quantize(timestamp - start[arg0, arg1]);
}
```

Reply:
I rolled that change in, and also for the perfgw.d script as I was updating it as well.